[lldb-dap] Fix running dap_server.py directly for debugging tests.

ashgti opened the pull request:
This adjusts the behavior of running dap_server.py directly to better support the current state of development. A few parts of the 'main' body were stale and not functional.

These improvements include:

* Instead of the custom tracefile / replay file parsing logic, I adjusted the replay helper to handle parsing lldb-dap log files created with the `LLDBDAP_LOG` env variable, allowing you to more easily run a failing test like: `python3 lldb/packages/Python/lldbsuite/test/tools/lldb-dap/dap_server.py --adapter lldb-dap -r lldb-test-build.noindex/tools/lldb-dap/console/TestDAP_console.test_custom_escape_prefix/dap.txt`
* Migrated argument parsing to `argparse`, that is in all verisons of py3+ and has a few improvements over `optparse`.
* Corrected the existing arguments and updated `run_vscode` > `run_adapter`. You can use this for simple debugging like: `xcrun python3 lldb/packages/Python/lldbsuite/test/tools/lldb-dap/dap_server.py --adapter=lldb-dap --adapter-arg='--pre-init-command' --adapter-arg 'help' --program a.out --init-command 'help'`

ashgti committed:
[lldb-dap] Fix running dap_server.py directly for debugging tests.

This adjusts the behavior of running dap_server.py directly to better support the current state of development.

* Instead of the custom tracefile parsing logic, I adjusted the replay helper to handle parsing lldb-dap log files created with the `LLDBDAP_LOG` env variable.
* Migrated argument parsing to `argparse`, that is in all verisons of py3+ and has a few improvements over `optparse`.
* Corrected the existing arguments and updated `run_vscode` > `run_adapter`. You can use this for simple debugging like: `xcrun python3 lldb/packages/Python/lldbsuite/test/tools/lldb-dap/dap_server.py --adapter=lldb-dap --adapter-arg='--pre-init-command' --adapter-arg 'help' --program a.out --init-command 'help'`

DavidSpickett reviewed (commented):
I have no context for the previous state of this, so I just looked at it generally. Someone else can be the approver.

ashgti committed:
Make sure we only split 1 time on '-->' or '<--' when replaying a log file.

da-viper commented on lldb/packages/Python/lldbsuite/test/tools/lldb-dap/dap_server.py:
```diff
@@ -1716,255 +1690,235 @@ def run_vscode(dbg, args, options):
-    if options.replay:
-        dbg.replay_packets(options.replay)
+    if opts.debug:
+        input('Waiting for debugger to attach pid "%i"' % (dbg.get_pid()))
```
```suggestion
        input(f'Waiting for debugger to attach pid "{dbg.get_pid()}"')
```

ashgti committed:
Re-write the replay method to only replay requests.

Additionally, I updated the replay logic to try to re-write some stateful values from requests/responses, otherwise we end up making requests with bad values like 'threadId' or 'frameId'.

ashgti commented on lldb/packages/Python/lldbsuite/test/tools/lldb-dap/dap_server.py:
```diff
@@ -1646,65 +1622,63 @@ def __str__(self):
+    source_to_lines: Dict[str, List[int]] = {}
+    for sbp in cast(List[str], opts.source_bp):
+        if ":" not in sbp:
+            print('error: invalid source with line "%s"' % (sbp))
```
Done.

da-viper commented on lldb/packages/Python/lldbsuite/test/tools/lldb-dap/dap_server.py:
```diff
@@ -1646,326 +1730,317 @@ def __str__(self):
     else:
-        if "message" in response:
-            print(response["message"])
+        print("failed to launch/attach: ", response)
```
```suggestion
        print("failed to launch/attach: ", response, file=sys.stderr)
```

ashgti committed:
Forwarding additional output to stderr.

ashgti committed:
Merge remote-tracking branch 'origin/main' into lldb-dap-test-debugging-helpers

ashgti committed:
Merge branch 'main' into lldb-dap-test-debugging-helpers
